File: toolfun/movieManagement/Package.java
// The first thing to do when a package control panel is opened is to check
// whether an object of the same class exists in the list of packages.
//
// * If the package already exists:
// - a pop up windows should notify the user that the previous package information
//   wil be used.
//
// * otherwise:
// - Call the constructor of the pacakge. Example:
//
//   newPackage = BiosensorPackage(movieData);
//
// In both case, call the sanity check. Any exception thrown during that check
// set the flag icons (not run, problem, dependency problem) next to each process.
//
//   try {
//     newPackage.sanityCheck(full);
//   } catch(ExceptionProcesses) {
//     // display warning icons according to exceptions thrown
//   } catch(otherTypeOfExceptions) {
//     // ...
//   } finally {
//     // No exception caught: every process has to have a "green check" icon
//   }
//	
// - If no critical error occured and the package was new, add the package to the
// movieData package list:
//
//   movieData.addPackage(newPackage);
//
// This class defines the abstract class Package from which every user-defined
// package will inherit.

public abstract class Package {
  // Constructor
  protected Package(movieData owner, String name)
  {
    owner_ = owner;
    name_ = name;
		
    // call getProcessDependenyMatrix() to get the size of the matrix.
    Matrixc depMatrix = getProcessDependenyMatrix_();    
    
    int nProcessses = size(depMatrix, 1);

    // initialize the cell array of processes
    // Matlab: processes_ = cell(nProcesses, 1);
    processes_ = new Process[nProcesses];

    // NEEDS TO BE OUTSIDE THE CTR:

    // the number of total processes stored in the movieData
    int nGlobalProcesses = length(owner_.processes_);
		
    // Check whether a required process has not been already computed by
    // another Package. If so, a pop window should ask the user whether he
    // wants to use the already computed process or he prefers to create a
    // new one. This pop up window should have 2 buttons 'Choose...'
    // and 'Create'. 'Choose' will open another window with the list of
    // existing related processes. 'Create' will create a new process with
    // default parameter.
		
    // Get the class name of require processes
    String[] processClassNames = getListProcessClassName();

    // Iterate over every required process and check whether processes of
    // the same class exists
    boolean isSame = new boolean[nGlobalProcesses];
    // isSame = false(nGlobalProcesses,1);

    for (int i = 0; i < nProcesses; ++i) {
			
      String className = processClassNames[i];

      for (int j = 0; j < nGlobalProcesses; ++j) {
	isSame[j] = classname(owner_.processes_[j]) == className;
      }
			
      if (any(isSame)) {
	// At least one process is of the same class as the required
	// ith process. This is when we need to show the popup window
	// with "choose..." and "create" button (see header at the top
	// of the file). Every owner_.processes_[j] where isSame[j] ==
	// true needs to be displayed as a potential choice.
				
	// TODO: ask user
				
	// test the user answer
	if (response == "choose..." && chosenProcess ~= end) {
	  // newPackage.setProcess(i, owner_.processes_[chosenProcess]);
	  processes_[i] = owner_.processes_[chosenProcess];
	}
      }
    }		
  }
	
  // Get the name of the package
  public String getName() {
    return name_;
  }
  
  // Get the number of processes
  public int getNumberOfProcesses() {
    return processes_.length();
  }

  // Set process
  public void setProcess(int iProcess, Process p) {
    assert(iProcess < size(depMatrix_, 1));

    processes_[iProcess] = p;
  }

  // make a sanity check of the package
  public abstract void sanityCheck(boolean full);
	
  // get the list of process class names required by the package.
  public abstract String[] getListProcessClassName();
	
  // This method return the matrix of process dependency. It represents the
  // directed graph of processes. Only the lower triangular part of the matrix
  // is considered. For every i > j, mat(i, j) == 1 means the ith process
  // depends on the jth process. If there is no dependency between processes,
  // return a zero matrix, where the size is equal to the number of processes.
  // Any process should have selft dependency, i.e. there should not be any
  // mat(i,i) == 1. The type of the matrix should be logical, i.e.
  // mat = false(nProcess); mat(i,j) = true, etc.
  public abstract Matrix getProcessDependenyMatrix();
	
  // This method traverses the list of processes and call for each process
  // its sanityCheck method. The list of processes is traversed using a depth
  // first search (DFS) algorithm to check dependency between processes.
  protected void checkProcesses_(boolean full) {
		
    // First, call sanityCheck method on each process and store any exception
    // in an array of exception.
    
    int nProcesses = processes_.length();

    // For each process, one needs to store every possible reasons a process
    // is wrong and display them onto the icon (ok, warning triangle, wrong)
    // next to the process.
    //
    // A process status can be alter for 3 reasons:
    // 1. the process itself has a problem
    // 2. the parameters in the process setting panel have changed
    // 3. the process depends on processes which also have a problem.
    //
    // We define an array of vector (Java). In Matlab it will be a cell array,
    // i.e.
    // processExceptions = cell(nProcesses, 1);
    //
    // If a process doesn't have any problem, the cell will be empty.
		
    Vector[] processExceptions = new Vector[nProcesses];
		
    // 1. Check whether the process itself has a problem
		
    for (int i = 0; i < nProcesses; ++i) {
      if (processes_[i]) {
	try {
	  processes_[i].sanityCheck(full);
	}
	catch (Exception) {
	  processExceptions[i].add(Exception);
	}
      }
    }
		
    // 2. Check whether the process parameters have changed
		
    for (int i = 0; i < nProcesses; ++i) {
      if (processes_[i] && processes_[i].hasChanged()) {
	newException = new Exception("Some of the process parameters have changed since previous run.");
	
	processExceptions[i].add(newException);
      }
    }
		
    // 3. Check if a process depends on processes which also have a problem.
		
    boolea[] processVisited = new false[nProcesses];
		
    for (int i = 0; i < nProcesses; ++i) {
      if (processes_[i] && !processVisited[i])
	dfs_(i, processExceptions, processVisited);
    }
		
    // Finally, if at least one of the process has a problem, we throw the
    // processExeptions cell array (the way it is done in Matlab is specific
    // so I just write something simple here: 
    if any(processExceptions) {
	throws processExceptions;
      }
  }
	
  private void dfs_(int iProcess, Vector[] processExceptions, processVisited) {
		
    processVisited[iProcess] = true;
		
    for (int j = 0; j < iProcess; ++j) {
      // does process iProcess depend on j ?
      if (depMatrix_[iProcess, j] == 1) {
	// ok, process iProcess depends on process j. Do we need to
	// explore process j?
	if (processes_[j] && !processVisited[j]) {
	  // yes
	  dfs_(j, processExceptions, processVisited);
	}

	// if jth process has a problem, we add an exception to the
	// ith process list of exception since ith process depends
	// on the jth process.
	if (processExceptions[j].length != 0 || !processes_[j]) {
	  newException = new Exception("Process " + iProcess +
				       " depends on process " +
				       j + "which has problem.");
					
	  processExceptions[iProcess].add(newException);
	}
      }
    }
  }
	
  // private fields
  private String name_;
	
  // protected fields section
  protected movieData owner_;
  // dependency matrix
  protected Matrix depMatrix_;
  // list of required processes
  protected Process[] processes_;
}

File: toolfun/movieManagement/Package.m
classdef Package < hgsetget 
    % Defines the abstract class Package from which every user-defined packages
    % will inherit. This class cannot be instantiated.
    
    properties(SetAccess = immutable)
        createTime_ % The time when object is created.
    end

    properties(SetAccess = protected)
        owner_ % The MovieData object this package belongs to
        processes_ % Cell array containing all processes who will be used in this package
        depMatrix_ % Processes' dependency matrix        
    end

    properties
        notes_ % The notes users put down
        outputDirectory_ %The parent directory where results will be stored.
                         %Individual processes will save their results to
                         %sub-directories of this directory.
    end

    methods
        function set.outputDirectory_(obj,value)
            
            if isequal(obj.outputDirectory_,value), return; end
            if ~isempty(obj.outputDirectory_), 
                stack = dbstack;
                if strcmp(stack(3).name,'MovieData.relocate'),
                    error(['This channel''s ' propertyName ' has been set previously and cannot be changed!']);
                end
            end
            endingFilesepToken = [regexptranslate('escape',filesep) '$'];
            value = regexprep(value,endingFilesepToken,'');
            obj.outputDirectory_=value;
        end
    end
    methods (Access = protected)
        function obj = Package(owner, depMatrix, outputDirectory,varargin)
            % Constructor of class Package
            
            if nargin > 0
                obj.owner_ = owner; 
                obj.depMatrix_ = depMatrix;
                obj.outputDirectory_ = outputDirectory;
                
                nVarargin = numel(varargin);
                if nVarargin > 1 && mod(nVarargin,2)==0
                    for i=1 : 2 : nVarargin-1
                        obj.(varargin{i}) = varargin{i+1};
                    end
                end
            
                obj.processes_ = cell(1,length(obj.getProcessClassNames));
                obj.createTime_ = clock;
            end
        end
        
        function [processExceptions, processVisited] = checkParentSanity(obj, ...
                procID,processExceptions,processVisited)
            % Check the dependencies sanity and return processExceptions
            
            processVisited(procID) = true;
            
            % Get the parent processes and remove empty optional processes
            parentIndex = obj.getParent(procID);
            if isempty(parentIndex), return;  end
            
            for parentID = parentIndex
                % Recursively call dependencies check for unvisited processes
                if ~isempty(obj.processes_{parentID}) && ~processVisited(parentID)
                    [processExceptions, processVisited] = ...
                        obj.checkParentSanity(parentID, processExceptions,processVisited);
                end
                
                % Check parent process validity and add exception if 
                % 1 - parent process is empty
                % 2 - parent process has at least one exception
                % 3 - required parent proc is not run successfully    
                isValidParent = @(x) ~isempty(x) && isempty(processExceptions{x}) && ...
                    obj.processes_{x}.success_;
                
                if obj.processes_{procID}.success_ && ~isValidParent(parentID)
                    % Set process's updated=false
                    obj.processes_{procID}.setUpdated (false);
                    
                    % Create a dependency error exception
                    statusMsg1 =['The step ' num2str(procID),': ' obj.processes_{procID}.getName...
                        ' is out of date because '];
                    statusMsg3 = '\nPlease run again to update your result.';
                    % Customized error message for first-time run optional processes
                    if obj.depMatrix_(procID,parentID)==2 && ~obj.processes_{parentID}.success_
                        statusMsg2 = ['the optional step ' num2str(parentID),...
                            ': ', eval([obj.getProcessClassNames{parentID} '.getName']),...
                        ', changes the input data of current step.'];
                    else
                        statusMsg2 = ['the step ' num2str(parentID),': ',...
                            eval([obj.getProcessClassNames{parentID} '.getName']),...
                        ', which the current step depends on, is out of date.'];
                    end
                    ME = MException('lccb:depe:warn', [statusMsg1 statusMsg2 statusMsg3]);
                    % Add dependency exception to the ith process
                    processExceptions{procID} = horzcat(processExceptions{procID}, ME);
                end
            end   
        end
        
    end
    methods (Access = public)
        
        function [status processExceptions] = sanityCheck(obj, varargin)
            % sanityCheck is called by package's sanitycheck. It returns
            % a cell array of exceptions. Keep in mind, make sure all process
            % objects of processes checked in the GUI exist before running
            % package sanitycheck. Otherwise, it will cause a runtime error
            % which is not caused by algorithm itself.
            %
            % The following steps will be checked in this function
            %   I. The process itself has a problem
            %   II. The parameters in the process setting panel have changed
            %   III. The process that current process depends on has a
            %      problem
            %
            % OUTPUT:
            %   processExceptions - a cell array with same length of
            % processes. It collects all the exceptions found in
            % sanity check. Exceptions of i th process will be saved in
            % processExceptions{i}
            %
            % INPUT:
            %   obj - package object
            %   full - true   check 1,2,3 steps
            %          false  check 2,3 steps
            %   procID - A. Numeric array: id of processes for sanitycheck
            %            B. String 'all': all processes will do
            %                                      sanity check
            %
            
            nProcesses = length(obj.getProcessClassNames);
            status = false(1,nProcesses);
            processExceptions = cell(1,nProcesses);
            processVisited = false(1,nProcesses);
            
            ip = inputParser;
            ip.CaseSensitive = false;
            ip.addRequired('obj');
            ip.addOptional('full',true, @(x) islogical(x));
            ip.addOptional('procID',1:nProcesses,@(x) (isvector(x) && ~any(x>nProcesses)) || strcmp(x,'all'));
            ip.parse(obj,varargin{:});
            
            full = ip.Results.full;
            procID = ip.Results.procID;
            if strcmp(procID,'all'), procID = 1:nProcesses;end
            
            validProc = procID(~cellfun(@isempty,obj.processes_(procID)));
            if full 
                % I: Check if the process itself has a problem
                %
                % 1. Process sanity check
                % 2. Input directory  
                for i = validProc
                    try
                        obj.processes_{i}.sanityCheck;
                    catch ME
                        % Add process exception to the ith process
                        processExceptions{i} = horzcat(processExceptions{i}, ME);
                    end
                end
            end
            
            % II: Determine the parameters are changed if satisfying the
            % following two conditions:
            % A. Process has been successfully run (obj.success_ = true)
            % B. Pamameters are changed (reported by uicontrols in setting
            % panel, and obj.procChanged_ field is 'true')
            changedProcesses = validProc(cellfun(@(x) x.success_ && x.procChanged_,obj.processes_(validProc)));
            for i = changedProcesses                    
                % Set process's updated=false
                obj.processes_{i}.setUpdated(false);
                % Create an dependency error exception
                ME = MException('lccb:paraChanged:warn',['The step ' num2str(i),': ' obj.processes_{i}.getName...
                        ' is out of date because the channels or parameters have been changed.']);
                % Add para exception to the ith process
                processExceptions{i} = horzcat(processExceptions{i}, ME);
            end
            
            % III: Check if the processes that current process depends
            % on have problems
            for i = validProc
                if ~processVisited(i)
                    [processExceptions, processVisited]= ...
                        obj.checkParentSanity(i, processExceptions, processVisited);
                end
            end
            
            % Return array of boolean
            saneProc = validProc(cellfun(@isempty,processExceptions(validProc)) &...
                cellfun(@(x) x.success_ && ~x.procChanged_,obj.processes_(validProc)));
            status(saneProc)=true;
            
        end
        
        function setDepMatrix(obj,row,col,value)
            % row and col could be array
            obj.depMatrix_(row, col) = value;
        end
            
        function setProcess(obj, i, newProcess)
            % set the i th process of obj.processes_ to newprocess
            % If newProcess = [ ], clear the process in package process
            % list
            assert(i<=length(obj.getProcessClassNames),...
                'UserDefined Error: i exceeds obj.processes length');
            if isa(newProcess,obj.getProcessClassNames{i}) ||...
                    isempty(newProcess)      
                obj.processes_{i} = newProcess;
            else
                error('User-defined: input should be Process object or empty.')
            end
        end
        
        function parentID = getParent(obj,procID)
            % Returns the list of valid parents for a given process
            
            % By default, get the required parent processes as well as
            % all non-empty optional parent processes
            reqParentIndex = find(obj.depMatrix_(procID,:)==1);
            optParentIndex = find(obj.depMatrix_(procID,:)==2);
            isValidOptParent = ~cellfun(@isempty,obj.processes_(optParentIndex));
            validOptParentIndex = optParentIndex(isValidOptParent);
            parentID=sort([reqParentIndex,validOptParentIndex]);
        end
        
    end
        
    methods(Static)
        function tools = getTools()
            tools=[];
        end
    end 

    methods(Static,Abstract)
        GUI
        getName
        getDependencyMatrix
        getOptionalProcessId
        getProcessClassNames
        getDefaultProcessConstructors
    end
    
end